### PR TITLE
New class for generating and loading data

### PR DESCRIPTION
**Pull request by prescod**
A new task to Orchestrate creating tempfiles, generating data, loading data, cleaning up tempfiles and batching.

I'll add some pytests after the initial review of the Task API.

### CODE REVIEW

File: cumulusci/tasks/bulkdata/generate_and_load_data.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        if not database_url:
+            sqlite_path = os.path.join(tempdir, f"generated_data_{index}.db")
+            database_url = f"sqlite:///" + sqlite_path
+        subtask_options = {
+            **self.options,
+            "mapping": mapping_file,
+            "database_url": database_url,
+            "num_records": batch_size,
+        }
+        self._datagen(subtask_options)
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/generate_and_load_data.py]: It seems like we should be passing the index into the data generator, in case it's important in order to generate a full sequence that spans multiple batches or something like that.

File: cumulusci/tasks/bulkdata/__init__.py [hunk 2]
```diff
@@ -1,7 +1,15 @@
...
 
 # For backwards-compatibility
 QueryData = ExtractData
+__all__ = (
+    "DeleteData",
+    "ExtractData",
+    "GenerateMapping",
+    "LoadData",
+    "GenerateAndLoadData",
+)
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/__init__.py]: It actually is more complicated than that. I spent like 20 minutes yesterday researching how to do that. Flake8 can only put comments on the same line as the stuff it refers to. Because the class and module names are so long, Black is turning this into a multi-line comment and preventing the Flake8 comments from working. The recommended fix in a VERY long and contentious thread I read is __all__.

[the remark this answers is not in the document]
Comment (davisagli) [reply cumulusci/tasks/bulkdata/__init__.py]: Okay, fair enough.

(I didn't know before that flake8 was smart enough to recognize the use as a string in `__all__`, so that's nice to learn.)

File: cumulusci/tasks/bulkdata/generate_and_load_data.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+from cumulusci.tasks.salesforce import BaseSalesforceApiTask
+from cumulusci.tasks.bulkdata import LoadData
+from cumulusci.utils import temporary_dir
+from cumulusci.core.config import TaskConfig
+from cumulusci.core.utils import import_global
+from cumulusci.core.exceptions import TaskOptionsError
+
+
+class GenerateAndLoadData(BaseSalesforceApiTask):
+    """ Orchestrate creating tempfiles, generating data, loading data, cleaning up tempfiles and batching."""
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: In the docstring or in the task_docs?

[the remark this answers is not in the document]
Comment (davisagli) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: Usually for tasks it should go in task_docs, which is the source when you run `cci task doc > docs/tasks.rst` to collate the task docs into the file that Sphinx builds. But I don't think this one will be included there since it's not configured in cumulusci.yml by default.

File: cumulusci/tasks/bulkdata/generate_and_load_data.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            "description": "How many records to generate. Precise calcuation depends on the generator.",
+            "required": True,
+        },
+        "batch_size": {
+            "description": "How many records to create and load at a time..",
+            "required": False,
+        },
+        "mapping": {"description": "A mapping YAML file to use", "required": True},
+        "data_generation_task": {
+            "description": "Fully qualified class path of a task to generate the data. Use cumulusci.tasks.bulkdata.factory_generator if you would like to use a Factory Module.",
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: Sorry yes, it was factored out of existence. I updated the doc in a3187bc4c9239a52966da261ec31a5b89adc0f00

File: cumulusci/tasks/bulkdata/generate_and_load_data.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            )
+
+        with temporary_dir() as tempdir:
+            for current_batch_size, index in self._batches(num_records, batch_size):
+                self._generate_batch(
+                    database_url, tempdir, mapping_file, current_batch_size, index
+                )
+
+    @staticmethod
+    def _batches(num_records, batch_size):
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: Done in e46d7339e72b430163a06ef9a4b8019ca3be7c40

File: cumulusci/tasks/bulkdata/generate_and_load_data.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        num_batches = (num_records // batch_size) + 1
+        for i in range(0, num_batches):
+            if i == num_batches - 1:  # last batch
+                batch_size = num_records - (batch_size * i)  # leftovers
+            if batch_size > 0:
+                yield batch_size, i
+
+    def _datagen(self, subtask_options):
+        class_path = self.options.get("data_generation_task", None)
+        task_class = import_global(class_path)
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: Fixed.

File: cumulusci/tasks/bulkdata/generate_and_load_data.py (hunk 1 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: https://github.com/SFDO-Tooling/CumulusCI/blob/f0eaab0c77ced04550e135113c2d06d4dbb50bcf/cumulusci/tasks/bulkdata/generate_and_load_data.py#L96

File: cumulusci/tasks/bulkdata/generate_and_load_data.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: Moved and required now.

File: cumulusci/tasks/bulkdata/generate_and_load_data.py (hunk 1 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: Okay, the index is now passed around pervasively.

File: cumulusci/tasks/bulkdata/generate_and_load_data.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        },
+        "database_url": {
+            "description": "A URL to store the database (defaults to a transient SQLite file)",
+            "required": "",
+        },
+    }
+
+    def _run_task(self):
+        mapping_file = os.path.abspath(self.options["mapping"])
+        assert os.path.exists(mapping_file), f"{mapping_file} cannot be found."
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/generate_and_load_data.py]: This should raise TaskOptionsError; assertion errors are reported with a traceback.

File: cumulusci/tasks/bulkdata/generate_and_load_data.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        database_url = self.options.get("database_url")
+        num_records = int(self.options["num_records"])
+        batch_size = int(self.options.get("batch_size", num_records))
+        class_path = self.options.get("data_generation_task")
+        self.data_generation_task = import_global(class_path)
+
+        if database_url and batch_size != num_records:
+            raise TaskOptionsError(
+                "You may not specify both `database_url` and `batch_size` options."
+            )
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/generate_and_load_data.py]: It would be best to do all the initialization of options in the _init_options method. That way it'll fail early while initializing a flow that uses the task instead of waiting until the task is running. (I know you're not using this as part of a flow in this case, but that's the general motivation for separating the initialization of options.)

File: cumulusci/tasks/bulkdata/tests/test_bulkdata.py [hunk 9]
```diff
@@ -1604,3 +1605,22 @@ def test_split_dependencies__interviews_for_cycles(self, prompt):
...
+
+        batches = list(generate_batches(num_records=3, batch_size=1))
+        assert batches == [(1, 0), (1, 1), (1, 2)]
+
+        batches = list(generate_batches(num_records=3, batch_size=3))
+        assert batches == [(3, 0)]
+
+    def test_batching_with_remainder(self):
+        batches = list(generate_batches(num_records=20, batch_size=7))
+        assert batches == [(7, 0), (7, 1), (6, 2)]
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/tests/test_bulkdata.py]: Move this to test_utils so it parallels the structure of the code being tested.

File: cumulusci/tasks/bulkdata/generate_and_load_data.py (hunk 7 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: https://github.com/SFDO-Tooling/CumulusCI/blob/529f491e088e1e9c2dd6044609e9590c0868bc63/cumulusci/tasks/bulkdata/generate_and_load_data.py#L63

File: cumulusci/tasks/bulkdata/generate_and_load_data.py (hunk 8 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/generate_and_load_data.py]: https://github.com/SFDO-Tooling/CumulusCI/blob/feature/generate-and-load-task/cumulusci/tasks/bulkdata/generate_and_load_data.py#L59-L74

File: cumulusci/tasks/bulkdata/tests/test_bulkdata.py (hunk 9 above)
Comment (prescod) [reply cumulusci/tasks/bulkdata/tests/test_bulkdata.py]: I had done that but forgotten to push it out.

https://github.com/SFDO-Tooling/CumulusCI/blob/529f491e088e1e9c2dd6044609e9590c0868bc63/cumulusci/tasks/bulkdata/tests/test_utils.py#L60-L76

**Comment by prescod** [comment]
I think that the test failures are unrelated.

File: cumulusci/tasks/bulkdata/tests/test_generate_and_load.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                GenerateAndLoadData,
+                {
+                    "options": {
+                        "num_records": 20,
+                        "mapping": mapping_file,
+                        "data_generation_task": "cumulusci.tasks.bulkdata.tests.dummy_data_factory.GenerateDummyData",
+                        "batch_size": 0,
+                    }
+                },
+            )
```
Comment (davisagli) [inline cumulusci/tasks/bulkdata/tests/test_generate_and_load.py]: What's the reason for defining init_task instead of just calling _make task inside a `with self.assertRaises(TaskOptionsError):` block?

Comment (prescod) [reply cumulusci/tasks/bulkdata/tests/test_generate_and_load.py]: I guess just a habit/pattern of setup and then assertions after setup.

In 1dcf06e2a I moved the options into a variable to avoid that but still use the context manager.

**Comment by prescod** [comment]
I wanted to capture the information about the faker hack but in my testing it didn't make a big enough difference that I would want it to be a default behaviour of CCI. So I added a note about it in   694f83a9b1d557. Obviously the impact it makes varies widely based on how much you depend on Faker and so far it didn't make much difference to my test cases. If I find it makes a big difference in future test cases then maybe I will want to make it an official "option" or default behaviour.

File: cumulusci/tasks/bulkdata/tests/dummy_data_factory.py [hunk 11]
```diff
@@ -5,29 +5,31 @@
 class GenerateDummyData(ModuleDataFactory):
     """Generate data based on test mapping.yml"""
 
-    def make_records(self, num_records, factories):
-        factories.create_batch("Contact", 10)
-        factories["Contact"].create_batch(5)
-        factories.create_batch("Household", 5)
+    def make_records(self, num_records, factories, current_batch_num):
+        assert num_records % 4 == 0, "Use a batch size divisible by 4"
```
[the remark this answers is not in the document]
Comment (prescod) [reply cumulusci/tasks/bulkdata/tests/dummy_data_factory.py]: Can you clarify/restate this comment? ELI5 it for me?

[the remark this answers is not in the document]
Comment (davidmreed) [reply cumulusci/tasks/bulkdata/tests/dummy_data_factory.py]: @prescod I apologize - I misinterpreted the parameters to this method. Please disregard!
